gh-149101: Implement PEP 788 (#149116)
python/cpython: Doc/c-api/interp-lifecycle.rst

:class:`KeyboardInterrupt`) and running the :mod:`atexit` functions, the runtime
is marked as finalizing, meaning that :c:func:`Py_IsFinalizing` and
:func:`sys.is_finalizing` return true.  At this point, only the finalization
thread (the thread that initiated finalization; this is typically the main thread)
is allowed to :term:`attach <attached thread state>` a thread state.

Other threads that attempt to attach during finalization, either explicitly
(such as via :c:func:`PyThreadState_Ensure` or :c:macro:`Py_END_ALLOW_THREADS`)
or implicitly (such as in-between bytecode instructions), will enter a
**permanently blocked state**. Generally, this is harmless, but this can
result in deadlocks. For example, a thread may be permanently blocked while
holding a lock, meaning that the finalization thread can never acquire that
lock.

Prior to CPython 3.13, the thread would exit instead of hanging,
which led to other issues (see the warning note at
:c:func:`PyThread_exit_thread`).

Gross? Yes. Starting in Python 3.15, there are a number of C APIs that make
it possible to avoid these issues by temporarily preventing finalization:

.. _interpreter-guards:

.. seealso::

>    :pep:`788`

Reviewer: Let's be clearer about why you'd want to read the historical document?
```suggestion
   :pep:`788` explains the design, motivation and rationale
   for these APIs.
```

.. c:type:: PyInterpreterGuard

   An opaque interpreter guard structure.

   By holding an interpreter guard, the caller can ensure that the interpreter
   will not finalize until the guard is closed (through
   :c:func:`PyInterpreterGuard_Close`).

   When a guard is held, a thread attempting to finalize the interpreter will

>    have to wait until the guard is closed before threads can be blocked.

Reviewer: Maybe use the term for “have to wait”?
```suggestion
   When a guard is held, a thread attempting to finalize the interpreter will
   block until the guard is closed before starting the finalization.
```

Author: Not opposed to the change, but I haven't been using "the" before "finalization" in other places. Was that intentional?

Reviewer: No!
```suggestion
   When a guard is held, a thread attempting to finalize the interpreter will
   block until the guard is closed before starting finalization.
```

   After finalization has started, threads are forever unable to acquire
   guards for that interpreter. This means that if you forget to close an
   interpreter guard, the process will **permanently hang** during
   finalization!

> 

Reviewer: Maybe compare/contrast the similar concept?
```suggestion

   Holding a guard for an interpreter is similar to holding a
   :term:`strong reference` to a Python object, except finalization does not happen
   automatically after all guards are released: it requires an explicit 
   :c:func:`Py_EndInterpreter` call.

```

   .. versionadded:: next

.. c:function:: PyInterpreterGuard *PyInterpreterGuard_FromCurrent(void)

   Create a finalization guard for the current interpreter. This will prevent

>    finalization from occuring until the guard is closed.

Reviewer: Cut the extra words?
```suggestion
   finalization until the guard is closed.
```

deleted. This can be done using interpreter views.

.. c:type:: PyInterpreterView

   An opaque view of an interpreter.

   This is a thread-safe way to access an interpreter that may have be
   finalizing or already destroyed.

   .. versionadded:: next

.. c:function:: PyInterpreterView *PyInterpreterView_FromCurrent(void)

   Create a view to the current interpreter.

   This function is generally meant to be used alongside
   :c:func:`PyInterpreterGuard_FromView` or :c:func:`PyThreadState_EnsureFromView`.

   On success, this function returns a view to the current interpreter; on
   failure, it returns ``NULL`` with an exception set.

   The caller must hold an :term:`attached thread state`.

   .. versionadded:: next

.. c:function:: void PyInterpreterView_Close(PyInterpreterView *view)

   Close an interpreter view.

   If an interpreter view is never closed, the view's memory will never be
   freed, but there are no other consequences. (In contrast, forgetting to
   close a guard will infinitely hang the main thread during finalization.)

   Currently, this function will deallocate *view*, but this may change in
   the future.

> 

Reviewer: Is this relevant for the user? Allocation looks like an implementation detail here.

Author: Not sure -- I added it thinking about debugging (e.g., if you're trying to find a missing `PyInterpreterGuard_Close` call, you can just look for unfreed `PyInterpreterGuard *` pointers).

Reviewer: If you're in a debugger already, I don't think you need docs to notice this implementation detail.

Author: Ok, I'll remove it.

   This function cannot fail, and the caller doesn't need to hold an
   :term:`attached thread state`.

.. c:function:: PyInterpreterView *PyInterpreterView_FromMain()

   Create a view for the main interpreter (the first and default
   interpreter in a Python process; see
   :c:func:`PyInterpreterState_Main`).

   On success, this function returns a view to the main
   interpreter; on failure, it returns ``NULL`` without an exception set.
   Failure indicates that the process is out of memory or that the main
   interpreter has finalized (or never existed).

   Generally speaking, using this function is strongly discouraged, because

>    it typically compromises subinterpreter support for a program. It exists

Reviewer: Cut the fluff?
```suggestion
   Using this function is strongly discouraged, because
   it typically compromises subinterpreter support for a program. It exists
```

Reviewer: Or be specific -- this is fine for embedders or top-level frameworks/apps?
```suggestion
   Using this function in extension libraries is strongly discouraged, because
   it typically compromises the library's subinterpreter support. It exists
```

deleted. This can be done using interpreter views.

.. c:type:: PyInterpreterView

   An opaque view of an interpreter.

   This is a thread-safe way to access an interpreter that may have be
   finalizing or already destroyed.

   .. versionadded:: next

.. c:function:: PyInterpreterView *PyInterpreterView_FromCurrent(void)

   Create a view to the current interpreter.

   This function is generally meant to be used alongside
   :c:func:`PyInterpreterGuard_FromView` or :c:func:`PyThreadState_EnsureFromView`.

   On success, this function returns a view to the current interpreter; on
   failure, it returns ``NULL`` with an exception set.

   The caller must hold an :term:`attached thread state`.

   .. versionadded:: next

.. c:function:: void PyInterpreterView_Close(PyInterpreterView *view)

   Close an interpreter view.

   If an interpreter view is never closed, the view's memory will never be
   freed, but there are no other consequences. (In contrast, forgetting to
   close a guard will infinitely hang the main thread during finalization.)

   This function cannot fail, and the caller doesn't need to hold an
   :term:`attached thread state`.

   .. versionadded:: next

> .. c:function:: PyInterpreterView *PyInterpreterView_FromMain()

Reviewer 2: ```suggestion
.. c:function:: PyInterpreterView *PyInterpreterView_FromMain(void)
```

   Create a view for the main interpreter (the first and default
   interpreter in a Python process; see
   :c:func:`PyInterpreterState_Main`).

   On success, this function returns a view to the main
   interpreter; on failure, it returns ``NULL`` without an exception set.
   Failure indicates that the process is out of memory or that the main

>    interpreter has finalized (or never existed).

Reviewer 2: This looks inaccurate: it only fails on OOM and returns a valid view when the interpreter is finalized or not yet initalized.

(Found by Claude)

>    Using this function in extension libraries is strongly discouraged, because

Reviewer 2: We should use an [affirmative tone](https://devguide.python.org/documentation/style-guide/#affirmative-tone). For example:

> Use this function when an interpreter pointer or view cannot be supplied by the caller, for example, when a native threading library does not provide a void *arg parameter that could carry a :c:type:`PyInterpreterGuard` or :c:type:`PyInterpreterView`. In code that supports subinterpreters, prefer :c:func:`PyInterpreterView_FromCurrent` so the guard tracks the calling interpreter rather than the main one.
